cm31_ntt/src/ntt.rs

```diff
@@ -0,0 +1,1289 @@
+use serde::{Serialize, Deserialize};
+use crate::cm31::CF;
+use num_traits::{Zero, One, Pow};
+use crate::ntt_utils::*;
+use anyhow::{Result, anyhow, ensure};
+
+pub const NTT_BLOCK_SIZE_FOR_CACHE: usize = 8usize.pow(5);
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct PrecomputedTwiddles {
+    pub small: Vec<CF>,  // for the base NTT of size NTT_BLOCK_SIZE_FOR_CACHE
+    pub full: Vec<CF>,   // for the higher levels of recursion in the radix-8 context
+    pub stride: Vec<CF>, // for stride 2 or stride 4
+}
+
+pub fn ntt(
+    f: &[CF],
+    precomp: &PrecomputedTwiddles,
+) -> Result<Vec<CF>> {
+    let n = f.len();
+    let mut scratch = vec![CF::zero(); NTT_BLOCK_SIZE_FOR_CACHE];
+    debug_assert!(is_power_of(n as u32, 2), "n must be a power of 2");
+    debug_assert!(n >= 8, "n must be at least 8");
```

Comment:
what do you think about replacing these with `anyhow::ensure` since the function is already fallible?

Reply:
and building on top of that, make other `ntt_...` functions fallible and bubble up the errors

Reply:
I love that suggestion. It would make things more consistent. I've made these changes in commit 5b2a8d9c